### PR TITLE
Add per-file overrides using field lists

### LINKED ISSUE #11: Refractor and add directive.
I have already started working on something related, but I'm opening an issue for discussion. 
My idea right now is to split everything into separate files as fit, so files for the following:
- file to handle getting the description/cleaning up the title
- file to handle inserting the tags
- file to handle the directive
Is what currently comes to mind.
As for the directive, it would have an override for the image, the description, and the type, plus additional metadata that fit certain types (For example, the `article` type, and its metadata (publish time, author, etc etc)

If anyone has any comments/ideas feel free to comment

### PR DESCRIPTION
**Pull request by ItayZiv**
Use field lists to implement per page overrides
Closes #11.
It seems like `__init__.py` might be due for a slight rewrite after this change.

### CODE REVIEW

**Comment by ItayZiv** [comment]
Probably should be done at this point, but it really does seem like `__init__.py` is getting a bit messy, and `test_options.py` might also be due for a cleanup to make all the tests more similar.

File: sphinxext/opengraph/__init__.py [hunk 1]
```diff
@@ -38,10 +40,17 @@ def get_tags(
...
 
     # Set length of description
     try:
-        desc_len = int(config["ogp_description_length"])
+        try:
+            desc_len = int(fields["ogp-description-length"])
+        except (ValueError, KeyError):
+            desc_len = int(config["ogp_description_length"])
     except ValueError:
         desc_len = DEFAULT_DESCRIPTION_LENGTH
```
Comment (TheTripleV) [inline sphinxext/opengraph/__init__.py:55]: This would be more readable as 1 try with 2 excepts.

File: sphinxext/opengraph/__init__.py [hunk 2]
```diff
@@ -50,15 +59,24 @@ def get_tags(
     title_excluding_html = get_title(context["title"], skip_html_tags=True)
 
     # Parse/walk doctree for metadata (tag/description)
-    description = get_description(doctree, desc_len, [title, title_excluding_html])
+    if "ogp-description" in fields:
+        description = fields["ogp-description"]
+    else:
+        description = get_description(doctree, desc_len, [title, title_excluding_html])
```
Comment (TheTripleV) [inline sphinxext/opengraph/__init__.py:65]: Refactor using `fields.get`

File: sphinxext/opengraph/__init__.py [hunk 3]
```diff
@@ -50,15 +59,24 @@ def get_tags(
...
 
     tags = "\n  "
 
     # title tag
-    tags += make_tag("og:title", title)
+    tags += make_tag(
+        "og:title", title if "ogp-title" not in fields else fields["ogp-title"]
+    )
```
Comment (TheTripleV) [inline sphinxext/opengraph/__init__.py:72]: Refactor using `fields.get`

File: sphinxext/opengraph/__init__.py [hunk 4]
```diff
@@ -50,15 +59,24 @@ def get_tags(
...
 
     # type tag
-    tags += make_tag("og:type", config["ogp_type"])
+    tags += make_tag(
+        "og:type",
+        config["ogp_type"] if "ogp-type" not in fields else fields["ogp-type"],
+    )
+
```
Comment (TheTripleV) [inline sphinxext/opengraph/__init__.py:79]: Refactor using `fields.get`

File: sphinxext/opengraph/__init__.py [hunk 5]
```diff
@@ -86,7 +104,11 @@ def get_tags(
     tags += make_tag("og:url", page_url)
 
     # site name tag
-    site_name = config["ogp_site_name"]
+    site_name = (
+        config["ogp_site_name"]
+        if "ogp-site-name" not in fields
+        else fields["ogp-site-name"]
+    )
```
Comment (TheTripleV) [inline sphinxext/opengraph/__init__.py:111]: Refactor using `fields.get`

File: sphinxext/opengraph/__init__.py [hunk 6]
```diff
@@ -96,9 +118,14 @@ def get_tags(
 
     # image tag
     # Get basic values from config
-    image_url = config["ogp_image"]
-    ogp_use_first_image = config["ogp_use_first_image"]
-    ogp_image_alt = config["ogp_image_alt"]
+    if "ogp-image" in fields:
+        image_url = fields["ogp-image"]
+        ogp_use_first_image = False
+        ogp_image_alt = fields["ogp-image-alt"] if "ogp-image-alt" in fields else None
```
Comment (TheTripleV) [inline sphinxext/opengraph/__init__.py:124]: Refactor using `fields.get`

**Review by TheTripleV — CHANGES_REQUESTED** [review]
Thanks for this PR. It is much needed. I didn't notice it was marked as ready till now, sorry!

The PR seems fine logic wise, there's just a few changes I want to increase readability of the code.

**Comment by TheTripleV** [comment]
Regarding the names of the fields, do field lists support underscores? If so, I think using underscores instead of hyphens would be better as the variable names in conf.py and the names of the field list keys would match.

Also, since making this extension support all opengraph tags would be nice. So, transforming all field list entries of the form
`:og:X: Y` to tags mapping `og:X` to `Y`.

**Comment by ItayZiv** [comment]
Regarding the underscores for names, I think it's supported but field lists everywhere else use hyphens, so I think it'd be better to be consistent with that.

File: sphinxext/opengraph/__init__.py (hunk 1 above)
Comment (ItayZiv) [reply sphinxext/opengraph/__init__.py:55]: Not sure I understand what you mean by 1 try with 2 excepts? Although it might go down to 1 try when I use `fields.get` anyways.. (Need to check if that works)

Comment (TheTripleV) [reply sphinxext/opengraph/__init__.py:55]: You can chain excepts like elifs

Comment (ItayZiv) [reply sphinxext/opengraph/__init__.py:55]: But a `ValueError` can be thrown inside both the inner try and the except, specifically, if they were chained excepts, and the config contained a non-integer value, a `ValueError` would be thrown.

File: README.md [hunk 7]
```diff
@@ -60,3 +60,40 @@ ogp_custom_meta_tags = [
...
+  * Sets the alt text. Will be ignored if there is no image set.
+
+### Example
+Remember that the fields **must** be placed at the very start of the file. You can verify Sphinx has picked up the fields if they aren't shown in the final html file.
+
+```rst
+:og:description: New description
+:og:image: http://example.org/image.png
+:og:image:alt: Example Image
+```
```
Comment (TheTripleV) [inline README.md:92]:
```suggestion
```rst
:og:description: New description
:og:image: http://example.org/image.png
:og:image:alt: Example Image

Page contents
=============
```
```

The examples should be fleshed out a little to show that the field list is at the top of the file.

File: README.md [hunk 8]
```diff
@@ -60,3 +60,40 @@ ogp_custom_meta_tags = [
...
+
+### Arbitrary Tags
+Additionally, you can use field lists to add any arbitrary OpenGraph tag not supported by the extension. The syntax for arbitrary tags is the same with `:og:tag: content`. For Example:
+
+```rst
+:og:video: http://example.org/video.mp4
+```
```
Comment (TheTripleV) [inline README.md:99]:
```suggestion
```rst
:og:video: http://example.org/video.mp4

Page contents
=============
```
```

same as above

File: sphinxext/opengraph/__init__.py [hunk 9]
```diff
@@ -29,6 +29,8 @@
 
 
 def make_tag(property: str, content: str) -> str:
+    # Parse quotation, so they won't break html tags if smart quotes are disabled
+    content = content.replace('"', "&quot;")
```
Comment (TheTripleV) [inline sphinxext/opengraph/__init__.py:33]: This should be in a separate PR.

File: sphinxext/opengraph/descriptionparser.py [hunk 10]
```diff
@@ -124,5 +124,4 @@ def get_description(
 
     mcv = DescriptionParser(description_length, known_titles, document)
     doctree.walkabout(mcv)
-    # Parse quotation so they won't break html tags if smart quotes are disabled
-    return mcv.description.replace('"', "&quot;")
+    return mcv.description
```
Comment (TheTripleV) [inline sphinxext/opengraph/descriptionparser.py:127]: This should be in the separate PR from above.

File: sphinxext/opengraph/__init__.py [hunk 11]
```diff
@@ -114,20 +129,24 @@ def get_tags(
...
-    tags += "\n".join(config["ogp_custom_meta_tags"])
+    # arbitrary tags and overrides
+    tags.update({k: v for k, v in fields.items() if k.startswith("og:")})
 
-    return tags
+    return (
+        "\n"
+        + "\n".join([make_tag(p, c) for p, c in tags.items()])
+        + "\n".join(config["ogp_custom_meta_tags"])
+    )
```
Comment (TheTripleV) [inline sphinxext/opengraph/__init__.py:149]:
```suggestion
    return (
        "\n"
        + "\n".join([make_tag(p, c) for p, c in tags.items()] + config["ogp_custom_meta_tags"])
    )
```

**Review by TheTripleV — CHANGES_REQUESTED** [review]

File: sphinxext/opengraph/__init__.py (hunk 9 above)
Comment (ItayZiv) [reply sphinxext/opengraph/__init__.py:33]: This was in a separate PR, it was moved so it would work with tags that come from field lists.

File: sphinxext/opengraph/descriptionparser.py (hunk 10 above)
Comment (ItayZiv) [reply sphinxext/opengraph/descriptionparser.py:127]: again, since this already was in a separate PR, it just gets moved so that it will also run on tags added via field lists, no real point in separating it since up until this PR the old location worked identically.

File: README.md [hunk 12]
```diff
@@ -28,7 +28,7 @@ Users hosting documentation on Read The Docs *do not* need to set any of the fol
 * `ogp_site_name`
     * This is not required. Name of the site. This is displayed above the title.
 * `ogp_image`
-    * This is not required. Link to image to show.
+    * This is not required. Link to image to show. Note that all relative paths are converted to be relative to the root of the html output.
```
Comment (Daltz333) [inline README.md:31]: We should add a comment to see an example below.

File: sphinxext/opengraph/__init__.py [hunk 13]
```diff
@@ -38,10 +40,17 @@ def get_tags(
     doctree: nodes.document,
     config: Dict[str, Any],
 ) -> str:
+    # Get field lists for per-poge overrides
```
Comment (Daltz333) [inline sphinxext/opengraph/__init__.py:43]:
```suggestion
    # Get field lists for per-page overrides
```

File: sphinxext/opengraph/descriptionparser.py (hunk 10 above)
Comment (Daltz333) [inline sphinxext/opengraph/descriptionparser.py:127]: Why was this changed?

File: README.md (hunk 12 above)
Comment (ItayZiv) [reply README.md:31]: what do you mean? there are no examples of the relative paths, I could just add another line with a example to the same sentence?

File: sphinxext/opengraph/descriptionparser.py (hunk 10 above)
Comment (ItayZiv) [reply sphinxext/opengraph/descriptionparser.py:127]: This wasn't removed but moved to __init__.py (make_tag function) so it would get applied on all the created tags, including the ones added via field lists.

**Review by Daltz333 — APPROVED** [review]

**Comment by TheTripleV** [comment]
Note, relative image paths in the custom overrides are broken (intentionally) in this PR. A fix should be included as a part of #53.
